### PR TITLE
melpa2nix: update to work with Emacs HEAD

### PR DESCRIPTION
**Pull request by purcell**
## Description of changes

When building recent Emacs HEAD with Nix, `melpaPackages` do not build, since the `package-build.el` version used here relies on a function in `package.el` that has been removed.

To resolve this, this PR makes `melpa2nix` use a newer version of `package-build` that has been fixed to work with Emacs HEAD. See https://github.com/melpa/package-build/pull/87

Some changes have also been necessary to the corresponding patch and to the elisp in `melpa2nix.el`.

_Edited for clarity_

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by purcell** [comment]
Running nixpkgs-review...

**Comment by purcell** [comment]
Seems to rebuild all the emacs packages okay on my `aarch64-darwin` machine, with the exception of those few with failing native portions (like elisp-ffi) which would not have been affected by this change.

Here's an example of the problem this fixes in the wild btw: https://github.com/purcell/nix-emacs-ci/actions/runs/7328520844/job/19956927982

**Review by tadfisher — APPROVED** [review]

**Comment by mweinelt** [comment]
22k rebuilds across four platforms onto master is not really okay. Even if the builds are small, they clog hydra for a long time, disrupting its capability to deliver timely build results to end users.

The rebuild flags really do reflect a traffic signal, and red means you have to stop.

**Comment by purcell** [comment]
Apologies, I've only opened PRs against `master` previously, but I think I understand the issue.

**Comment by AndersonTorres** [comment]
> Apologies, I've only opened PRs against `master` previously, but I think I understand the issue.

Don't feel so bad. This was a recent unwritten decision.

**Comment by delroth** [comment]
> > Apologies, I've only opened PRs against `master` previously, but I think I understand the issue.
> 
> Don't feel so bad. This was a recent unwritten decision.

CONTRIBUTING.md linked in every single PR says: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#changes-causing-mass-rebuilds

> Furthermore, if the change causes a mass rebuild, use the appropriate staging branch instead 

> As a rule of thumb, if the number of rebuilds is over 500, it can be considered a mass rebuild.

There is nothing recent about this, it's been the policy for years. While I also don't blame @purcell for this (they didn't press the merge button, and trying to catch these issues is why we do reviews), it's surprising to me that an ex-committer to nixpkgs would claim this is a "recent unwritten decision".

**Comment by purcell** [comment]
Thanks for sharing the docs. I did find that info after the fact.

I wonder if there's some stricter automation that could prevent this sort of issue, e.g.

* Make ofborg add a more explicit "merge-prohibited" label that's hard to miss (bright red is not perceived the same by everyone)
* Produce a failing Actions check on the basis of the target branch and projected number of rebuilds
* Branch protection rules tied to one of the above, perhaps?

Effort spent on that automation might yield diminishing returns if the problems they mitigate are infrequent, I guess.

**Comment by vcunat** [comment]
"recent": for some period of time, emacs rebuilds were being merged directly into master with the argument that each of the many builds is very cheap.

**Comment by adisbladis** [comment]
I could have sworn I checked that this was targeted against staging before merging. That was a mistake.

**Comment by vcunat** [comment]
Moved to `staging-next` branch right now, as a compromise.

**Comment by AndersonTorres** [comment]
> > > Apologies, I've only opened PRs against `master` previously, but I think I understand the issue.
> > 
> > 
> > Don't feel so bad. This was a recent unwritten decision.
> 
> CONTRIBUTING.md linked in every single PR says: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#changes-causing-mass-rebuilds
> 
> > Furthermore, if the change causes a mass rebuild, use the appropriate staging branch instead

Usually, mass rebuilds were not about the raw counting of packages being rebuilt, but about the depth of dependency tree/graph.
Softwares like Meson, Cmake or even GCC are indirect dependencies of a huge amount of different and otherwise unrelated packages. It is understandable that their mass rebuilds should be treated more carefully, since they bring profound modifications on the system as a whole.

The same can't be said of Emacs. Emacs is mostly a dependency of Emacs Lisp packages, and Emacs Lisp packages are small and not useful outside Emacs itself.
Further, we have such a huge amount of Elisp packages because their inclusion into Nixpkgs is automated. Indeed we indulge ouserlves in relying on Emacs Overlay Project to provide up-to-date versions of those Emacs Lisp packages.

Also, Emacs is not a "plumbing" package. Usually no one except developers care about if their Cmake or Meson packages are up-to-date. The same can't be said about Emacs since it is a user package.

Because of all of those things, it was considered OK to merge Emacs changes over master.

> There is nothing recent about this, it's been the policy for years. 

The unwritten policy of merging Emacs changes over master is almost as old as, well, adisbladis contributions I suppose?

For a long amount of time, before the decision of providing precompiled Elisp packages, Emacs updates were sent to master branch. This trend continued for a very long time after that decision - indeed, I myself refactored the whole expression that builds two mainline, one experimental and two Macport releases, and absolutely no one thrown the Contributing Guide complaining about mass rebuilds.

> it's surprising to me that an ex-committer to nixpkgs would claim this is a "recent unwritten decision".

Maybe because you wasn't there?

https://github.com/NixOS/nixpkgs/pull/190706 - first time Elisp updates triggered red flags

https://github.com/NixOS/nixpkgs/pull/233301
https://github.com/NixOS/nixpkgs/pull/235859
https://github.com/NixOS/nixpkgs/pull/237125
https://github.com/NixOS/nixpkgs/pull/251472

**Comment by purcell** [comment]
The commit where this was re-applied in `staging-next` is 895cd4b126fc, btw. Hoping it lands soon... I couldn't find any way to judge progress or timelines.

**Comment by vcunat** [comment]
It should get into channels over the weekend.

**Pull request merged**
